app.py

```diff
@@ -45,35 +45,73 @@
 # Initialize rate limiting variables
 request_count = 0
 rate_limit_start = time.time()
-
+port_access_token, token_expiry_time, port_headers = None, datetime.now(), {}
 bitbucket_auth = BasicAuth(username=BITBUCKET_USERNAME, password=BITBUCKET_PASSWORD)
-
-# Obtain the access token synchronously
-credentials = {"clientId": PORT_CLIENT_ID, "clientSecret": PORT_CLIENT_SECRET}
-token_response = httpx.post(f"{PORT_API_URL}/auth/access_token", json=credentials)
-port_headers = {"Authorization": f"Bearer {token_response.json()['accessToken']}"}
-
-# Initialize the global AsyncClient with a timeout
 client = httpx.AsyncClient(timeout=httpx.Timeout(60))
 
-async def get_or_create_port_webhook():
-    logger.info("Checking if a Bitbucket webhook is configured on Port...")
+async def get_access_token():
+    credentials = {"clientId": PORT_CLIENT_ID, "clientSecret": PORT_CLIENT_SECRET}
+    token_response = await client.post(f"{PORT_API_URL}/auth/access_token", json=credentials)
+    response_data = token_response.json()
+    access_token = response_data['accessToken']
+    expires_in = response_data['expiresIn']
+    token_expiry_time = datetime.now() + timedelta(seconds=expires_in)
+    return access_token, token_expiry_time
+
+async def refresh_token_if_expired():
+    global port_access_token, token_expiry_time, port_headers
+    if datetime.now() >= token_expiry_time:
+        logger.info("Access token has expired, requesting a new one...")
+        port_access_token, token_expiry_time = await get_access_token()
+        port_headers = {"Authorization": f"Bearer {port_access_token}"}
+        logger.info(f"New token received. Expiry time: {token_expiry_time}")
```

Comment:
let's make this refresh_token_and_retry()

async refresh_token_and_retry(method, url, **kwargs)
  new_token = get_access_token()
  global port_headers  # Update global token
  port_headers = {"Authorization": f"Bearer {new_token}"}
   response = await self.client.request(method, url, headers=port_headers, **kwargs)
  return response

Reply:
Updated